File: cmdline/make_issue.pl
#!/usr/bin/perl
#####
# Licensed to the Apache Software Foundation (ASF) under one or more
# contributor license agreements.  See the NOTICE file distributed with
# this work for additional information regarding copyright ownership.
# The ASF licenses this file to You under the Apache License, Version 2.0
# (the "License"); you may not use this file except in compliance with
# the License.  You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#####
# make_issue
# A program for creating issues to be voted upon by the named "group"
#
# o  must be run by voter user (see wrapsuid.c for setuid wrapper)
#
# o  creates an issue directory (/home/voter/issues/group/YYYYMMDD-name/),
#    and fills it with files for the issue info, election monitors,
#    and tally file;
#
# o  creates tables of hash-ids and hashed-hash-ids for use in validating
#    votes, verifying that the values are unique;
#
# o  mails to the vote monitors an alert message to start tallying;
#
# o  mails to each voter their hash-id to be used, issue number, and
#    some text explaining the issue.
#
# Originally created by Roy Fielding
#
BEGIN {
    unshift @INC, "/home/voter/bin";
}
require "getopts.pl";
use randomize;
use ballots;
use steve;

umask(0077);
$| = 1;                                     # Make STDOUT unbuffered

$pname = $0;                                # executable name for errors
$pname =~ s#^.*/##;

# ==========================================================================
# ==========================================================================
# Print the usage information if help requested (-h) or a bad option given.
#
sub usage
{
    die <<"EndUsage";
usage: $pname [-h] [-b] [-g group] [-s start_date] [-i issue] [-f infofile]
                    [-m monitors] [-v vote_type )]
$pname -- Make an issue for managing an on-line, anonymous voting process
Options:
     -h  (help) -- just display this message and quit
     -b  (batch) -- batch processing: assume 'ok' unless errors
     -g  group  -- create an issue for an existing group of voters
     -s  date   -- YYYYMMDD format of date that voting is allowed to start
     -i  issue  -- append this alphanumeric string to start date as issue name
     -f  file   -- send the contents of this file to each voter as explanation
     -m  monitors -- e-mail address(es) for sending mail to vote monitor(s)
     -v  type   -- type of vote: yna = Yes/No/Abstain,
                   stvN = single transferable vote for [1-9] slots,
                   selectN = vote for [1-9] of the candidates

EndUsage
}

# ==========================================================================
# Get the command-line options or input from the user

undef $opt_h;
undef $opt_b;

if (!(&Getopts('hbg:s:i:f:m:v:')) || defined($opt_h)) { &usage; }

if (defined($opt_g)) {
    $group = $opt_g;
}
else {
    $group = &get_input_line("group name for voters on this issue", 1);
}
if ($group !~ /^\w+$/) {
    die "$pname: group name must be an alphanumeric token\n";
}

if (defined($opt_s)) {
    $start_date = $opt_s;
}
else {
    $start_date = &get_input_line("YYYYMMDD date that voting starts", 1);
}
if ($start_date !~ /^[2-9]\d\d\d(0[1-9]|1[012])([012]\d|3[01])$/) {
    die "$pname: start date must be formatted as YYYYMMDD, like 20020930\n";
}

if (defined($opt_i)) {
    $issue = $opt_i;
}
else {
    $issue = &get_input_line("short issue name to append to date", 1);
}
if ($issue !~ /^\w+$/) {
    die "$pname: issue name must be an alphanumeric token\n";
}

if (defined($opt_f)) {
    $infofile = $opt_f;
}
else {
    $infofile = &get_input_line("file pathname of issue info on $host", 1);
}
if (!(-e $infofile)) {
    die "$pname: info file does not exist: $infofile\n";
}
if ($infofile =~ /(\/etc\/|$issuedir)/) {
    die "$pname: forbidden to read info files in that directory\n";
}

if (defined($opt_m)) {
    $monitors = $opt_m;
}
else {
    $monitors = &get_input_line("e-mail address(es) for vote monitors", 1);
}
if ($monitors !~ /\@/) {
    die "$pname: vote monitor must be an Internet e-mail address\n";
}

if (defined($opt_v)) {
    $vote_type = $opt_v;
}
else {
    $vote_type = &get_input_line("vote type: yna, stvN, or selectN (N=1-9)", 1);
}
if ($vote_type =~ /^yna$/i) {
    $selector = 0;
    $style = "yes, no, or abstain";
}
elsif ($vote_type =~ /^stv([1-9])$/i) {
    $selector = $1;
    $style = "single transferable vote for $selector slots";
}
elsif ($vote_type =~ /^select([1-9])$/i) {
    $selector = $1;
    $style = "select $selector of the candidates labeled [a-z0-9]";
}
else {
    die "$pname: vote type must be yna, stvN, or selectN (N=[1-9])\n";
}

if (defined($opt_b)) {
    $batch = 1;
}
else {
    $batch = 0;
}

# ==========================================================================
# Check the voter group and read the list of voter addresses

die "$pname: cannot find $issuedir\n" unless (-d $issuedir);
$issuedir .= "/$group";
if (-d $issuedir) {
    die "$pname: you lack permissions on $issuedir\n"
        unless (-o _ && -r _ && -w _ && -x _);
}
else {
    die "$pname: group $group has not been created yet, see votegroup\n";
}

$votersfile = "$issuedir/voters";
@voters = &get_group($votersfile);
if ($#voters < 0) {
    die "$pname: $group must be an existing voter group: see votegroup\n";
}

# ==========================================================================
# Create directory for new issue only if it doesn't exist

$issuename = "$group-$start_date-$issue";
print "Creating new issue: $issuename\n";

$issuedir .= "/$start_date-$issue";
if (-e $issuedir) { die "$pname: $issuedir already exists\n"; }
mkdir($issuedir, 0700) || die "$pname: cannot mkdir $issuedir: $!\n";

# ==========================================================================
# Create issue information file (needs to be done before voter hash)

$issueaddr = 'voter@apache.org';

$issuefile = "$issuedir/issue";

open(ISF, ">$issuefile") || die "$pname: cannot open issue file: $!\n";
print ISF <<"EndOutput";
Hello Apache $group,

A call for votes has been declared for the following:

   Issue: $issuename
   Voting style: $style

EndOutput

open(INFILE, $infofile) || die "$pname: cannot open info file: $!\n";
print ISF <INFILE>;
close(INFILE);
close(ISF);

# ==========================================================================
# Create a table of voter hash-ids and verify uniqueness

for (;;) {
    $issid = &filestuff($issuefile);
    $monhash = &get_hash_of("$issid:$monitors");
    foreach $voter (@voters) {
        $h1 = &get_hash_of("$issid:$voter");
        $h2 = &get_hash_of("$issid:$h1");
        $hash1{$voter} = $h1;
        $hash2{$voter} = $h2;
        $invert1{$h1} = $voter;
        $invert2{$h2} = $voter;
    }
    $numh1 = scalar(keys(%hash1));
    $numh2 = scalar(keys(%hash2));
    $numi1 = scalar(keys(%invert1));
    $numi2 = scalar(keys(%invert2));
    last if (($numh1 == $numi1) && ($numh2 == $numi2));
    $h1 = &get_input_line("anything to retry collision-free hash", 1);
    system($TOUCH, $issuefile);
}
# &debug_hash;

# ==========================================================================
# Write list of monitors to issuedir/monitors
$monfile = "$issuedir/monitors";

open(MON, ">$monfile") || die "$pname: cannot open monitors file: $!\n";
print MON "$monitors\n";
close(MON);

# ==========================================================================
# Write type of vote to issuedir/vote_type
$typefile = "$issuedir/vote_type";

open(MON, ">$typefile") || die "$pname: cannot open vote type file: $!\n";
print MON "$vote_type\n";
if ($selector == 0) {
    @ballots = ( "yes\n", "no\n", "abstain\n");
} else {
    open(INFILE, $issuefile) || die "$pname: cannot open issue file: $!\n";
    @ballots = ballots(<INFILE>);
    close(INFILE);
}
print MON @ballots;
close(MON);

# ==========================================================================
# Verify with user that the info file is okay before mailing everyone.

print "Here is the issue information to be sent to each voter:\n";
print "==============================================================\n";
system($CAT, $issuefile);
&explain_vote(*STDOUT, 'unique-hash-key');
print "==============================================================\n";
do {
    $_ = ($batch ? "ok" : &get_input_line('"ok" to accept or "abort" to delete issue', 0));
    if (/^abort$/i) {
        system('rm', '-rf', $issuedir);
        exit(1);
    }
} until (/^ok/i);

# ==========================================================================
# Verify with user that the voters file is okay before mailing everyone.

print "Here is the list of voter e-mail addresses:\n";
print "==============================================================\n";
for $voter (@voters) {
    print $voter, "\n";
}
print "==============================================================\n";
do {
    $_ = ($batch ? "ok" : &get_input_line('"ok" to accept or "abort" to delete issue', 0));
    if (/^abort$/i) {
        system('rm', '-rf', $issuedir);
        exit(1);
    }
} until (/^ok/i);

# ==========================================================================
# Send mail to monitors telling them that the issue has been put to
# a vote, including the list of valid hashed hash-ids, sigs of files,
# and the issue info file.

open (MAIL, "|$SENDMAIL -t -f$issueaddr") || die("cannot send mail: $!\n");

print MAIL <<"EndOutput";
From: "Apache voting tool" <$issueaddr>
To: $monitors
Subject: Monitoring vote on $issuename

You have been selected as a vote monitor for Apache

   Issue: $issuename
   Voting style: $style

EndOutput
open(INFILE, $infofile) || die "$pname: cannot open info file: $!\n";
print MAIL <INFILE>;
close(INFILE);

print MAIL <<"EndOutput";

= = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = =

The voting system will record a tally of the votes as they are received
and will report that to you when the voting is closed.  To close voting,
use ssh to login to $host and then run

   /home/voter/bin/close_issue $issuename $monhash

For verification purposes, you will be receiving an e-mail notification
of each vote submitted by the voters.  Repeat votes should be considered
a complete replacement of the person's prior vote.  Your primary role in
all of this is to compare the votes you received with the results that
are tallied automatically, letting the group know if there is a
significant difference.  Any change made to the issue files during the
voting process will immediately invalidate the hash IDs and kill voting.

The following $numh2 voters are valid, as identified by the double-hashed ID
that will be sent to you when a vote is recorded.  Note that these IDs are
different from the single-hash IDs used by the voters when voting.

EndOutput

foreach $voter (sort(values(%hash2))) {
    print MAIL "   voter: $voter\n";
}

# Collect hash signatures of voter files that should not change
@vfiles = (
    "$votersfile",
    "$issuefile",
    "$monfile",
    "$typefile",
    "$homedir/bin/make_issue",
    "$homedir/bin/make_issue.pl",
    "$homedir/bin/vote",
    "$homedir/bin/vote.pl",
    "$homedir/bin/close_issue",
    "$homedir/bin/close_issue.pl"
);

print MAIL "\nThe following explains the voting process to voters:\n";
&explain_vote(*MAIL, 'unique-hash-key');
print MAIL "\nCurrent file digests:\n\n";
foreach $vf (@vfiles) {
    $pf = $vf;
    $pf =~ s/^$homedir\///o;
    print(MAIL &hash_file($vf), ': ', $pf, "\n");
}
close(MAIL);

# ==========================================================================
# Touch tally file to allow voting to begin
# -- Theoretically, this could be done by a another program with "at"
#    scheduling to prevent people from voting too early, but why bother?

system($TOUCH, "$issuedir/tally");

# ==========================================================================
# Send mail to voters telling them that the issue has been put to vote,
# including the info file and their commands.

while (($voter, $h1) = each %hash1) {
    print "Sending mail to voter: $voter\n";
########################################### for debugging
#   print "                hash1: $h1\n";
#   open (MAIL, ">>debug.txt") ||
########################################### replace next line
    open (MAIL, "|$SENDMAIL -t -f$issueaddr") ||
        die("cannot send mail to $voter: $!\n");

    print MAIL <<"EndOutput";
From: "Apache voting tool" <$issueaddr>
To: $voter
Subject: Apache vote on $issuename
Reply-To: $monitors

EndOutput
    open(INFILE, $issuefile) || die "$pname: cannot open issue file: $!\n";
    print MAIL randomize(<INFILE>);
    close(INFILE);
    &explain_vote(*MAIL, $h1);
    close(MAIL);
}

# ==========================================================================
print "Issue $issuename with hashcode of $monhash\nhas been successfully created.\n";
exit(0);


# ==========================================================================
# ==========================================================================
sub explain_vote {
    local (*FDES, $hashid) = @_;

    print FDES <<"EndOut1";

= = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = =

Your voting key for this issue: $hashid

In order to vote, either visit

   https://vote.apache.org/cast/$issuename/$hashid

or use ssh to login to $host and then run

   /home/voter/bin/vote $issuename $hashid "vote"

EndOut1

    if ($selector == 0) {
        print FDES <<"EndYNA";
where "vote" must be replaced by "yes", "no", or "abstain".
EndYNA
    }
    elsif ($vote_type =~ /^stv/i) {
        print FDES <<"EndSTV";
where "vote" must be replaced by a single word containing the
concatenated labels of candidates in the order that you wish them
to be selected.  In other words, if you want to vote for the candidates
labeled [x], [s], and [p], in that order, then your vote should be "xsp".

This election will be decided according to the Single Transferable Vote
rules described at

   http://wiki.apache.org/general/BoardVoting
   http://www.electoral-reform.org.uk/votingsystems/stvi.htm
   http://www.cix.co.uk/~rosenstiel/stvrules/index.htm

for an election with $selector open slots. 

You have one vote.  Use your vote by entering the label of your
first preference candidate followed by, if desired, the label of your
second preference candidate, and so on until you are indifferent about
the remaining candidates.  The sequence of your preferences is crucial.
You should continue to express preferences only as long as you are able
to place successive candidates in order.  A later preference is considered
only if an earlier preference has a surplus above the quota required for
election, or is excluded because of insufficient support.  Under no
circumstances will a later preference count against an earlier preference.

You may list as many candidates as you wish, but no more than once per
vote (e.g., "xsxp" would be rejected).
EndSTV
    }
    else {
        print FDES <<"EndSelect";
where "vote" must be replaced by a single word containing the
concatenated labels of your $selector choices.  In other words,
if you want to vote for the candidates labeled [x], [s], and [p],
then your vote should be "xsp" (order does not matter).
EndSelect
    }

    print FDES <<"EndExplain";
If for some reason you are unable to use ssh to access $host,
then you can vote by proxy: simply send your voting key to some
person with ssh access that you trust, preferably with instructions
on how you wish them to place your vote.

For verification purposes, you will be receiving an e-mail notification
each time your voting key is used.  Repeat votes will be considered
a complete replacement of your prior vote.  Your vote will be
recorded in a tally file and sent to the vote monitors along with
a different unique key, minimizing the chance that the contents of
your vote will be accidentally seen by someone else while associated
to you.  That is why the verification e-mail will only state that you
have voted, rather than including how you voted.

If you have any problems or questions, send a reply to the vote monitors
for this issue: $monitors

EndExplain
}

File: cmdline/make_issue.py
#!/usr/bin/env python
#
# Licensed to the Apache Software Foundation (ASF) under one or more
# contributor license agreements.  See the NOTICE file distributed with
# this work for additional information regarding copyright ownership.
# The ASF licenses this file to You under the Apache License, Version 2.0
# (the "License"); you may not use this file except in compliance with
# the License.  You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
#
# make_issue
#
# A program for creating issues to be voted upon by the named "group"
#
# o  must be run by voter user (see wrapsuid.c for setuid wrapper)
#
# o  creates an issue directory (/home/voter/issues/group/YYYYMMDD-name/),
#    and fills it with files for the issue info, election monitors,
#    and tally file;
#
# o  creates tables of hash-ids and hashed-hash-ids for use in validating
#    votes, verifying that the values are unique;
#
# o  mails to the vote monitors an alert message to start tallying;
#
# o  mails to each voter their hash-id to be used, issue number, and
#    some text explaining the issue.
#

import sys
import os
import argparse
import re
import shutil
import cStringIO

### how do we want to "properly" adjust path?
sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), '../lib')))
import steve
import ezt


def main():
  args = parse_argv()

  ### where should we look for this config file?
  config = steve.load_config('steve.conf')

  # Expand the set of arguments (as a side-effect), if they were not provided
  # on the cmdline.
  augment_args(args, config)

  issue_name = '%s-%s-%s' % (args.group, args.start, args.issue)

  # Note: config.issue_dir updated as a side-effect
  voters = get_voters(args, config)

  # Note: config.issue_dir updated as a side-effect
  create_issue_dir(issue_name, args, config)

  info_fname = create_info_file(issue_name, args, config)
  monitors_hash, hash = build_hash(info_fname, voters, args)

  monitors_fname = create_monitors_file(args, config)
  type_fname = create_type_file(info_fname, args, config)
  verify_email(issue_name, info_fname, args, config)
  verify_voters(voters, args, config)

  email_monitors(issue_name, info_fname, monitors_hash, hash, args, config)
  email_voters()

  print "Issue %s with hashcode of %s\nhas been successfully created." \
        % (issue_name, monitors_hash)


def parse_argv():
  parser = argparse.ArgumentParser(
             prog=steve.PROG,
             description='Make an issue for managing an on-line, '
                         'anonymous voting process',
             )
  parser.add_argument('-b', '--batch', action='store_true',
                      help='batch processing: assume "ok" unless errors')
  parser.add_argument('-g', '--group',
                      help='create an issue for an existing group of voters')
  parser.add_argument('-s', '--start',
                      help='YYYYMMDD format of date that voting is allowed to start')
  parser.add_argument('-i', '--issue',
                      help='append this alphanumeric string to start date '
                           'as issue name')
  parser.add_argument('-f', '--file',
                      help='send the contents of this file to each voter '
                           'as explanation')
  parser.add_argument('-m', '--monitors',
                      help='e-mail address(es) for sending mail to vote monitor(s)')
  parser.add_argument('-v', '--votetype',
                      help='type of vote: yna = Yes/No/Abstain, '
                           'stvN = single transferable vote for [1-9] slots, '
                           'selectN = vote for [1-9] of the candidates')

  return parser.parse_args()


def augment_args(args, config):
  "Update ARGS in-place with missing values."

  if args.group is None:
    args.group = steve.get_input_line('group name for voters on this issue', True)
  if not re.match(r'^\w+$', args.group):
    steve.die('group name must be an alphanumeric token')

  if args.start is None:
    args.start = steve.get_input_line('YYYYMMDD date that voting starts', True)
  if not re.match(r'^[2-9]\d\d\d(0[1-9]|1[012])(0[1-9]|[12][0-9]|3[01])$',
                  args.start):
    steve.die('start date must be formatted as YYYYMMDD, like 20090930')

  if args.issue is None:
    args.issue = steve.get_input_line('short issue name to append to date', True)
  if not re.match(r'^\w+$', args.issue):
    steve.die('issue name must be an alphanumeric token')

  if args.file is None:
    args.file = steve.get_input_line('file pathname of issue info on %s'
                                     % (config.hostname,),
                                     True)
  args.file = os.path.realpath(args.file)
  if not os.path.exists(args.file):
    steve.die('info file does not exist: %s', args.file)
  if args.file.startswith('/etc/'):
    steve.die('forbidden to read info files from: /etc')
  issue_dir = os.path.realpath(config.issue_dir)
  if args.file.startswith(issue_dir + '/'):
    steve.die('forbidden to read info files from: %s', issue_dir)

  if args.monitors is None:
    args.monitors = steve.get_input_line('e-mail address(es) for vote monitors', True)
  if '@' not in args.monitors:
    steve.die('vote monitor must be an Internet e-mail address')

  if args.votetype is None:
    args.votetype = steve.get_input_line('vote type; yna, stvN, or selectN (N=1-9)',
                                         True)
  args.votetype = args.votetype.lower()
  if args.votetype == 'yna':
    args.selector = 0
    args.style = 'yes, no, or abstain'
  elif args.votetype.startswith('stv') and args.votetype[3:].isdigit():
    args.selector = int(args.votetype[3:])
    args.style = 'single transferable vote for %d slots' % (args.selector,)
  elif args.votetype.startswith('select') and args.votetype[6:].isdigit():
    args.selector = int(args.votetype[6:])
    args.style = 'select %d of the candidates labeled [a-z0-9]' % (args.selector,)
  else:
    steve.die('vote type must be yna, stvN, or selectN (N=[1-9])')


def get_voters(args, config):
  if not os.path.isdir(config.issue_dir):
    steve.die('cannot find: %s', config.issue_dir)

  config.issue_dir += '/' + args.group
  if not os.path.isdir(config.issue_dir):
    steve.die('group "%s" has not been created yet, see votegroup', args.group)
  if not os.access(config.issue_dir, os.R_OK | os.W_OK | os.X_OK):
    steve.die('you lack permissions on: %s', config.issue_dir)
  uid = os.stat(config.issue_dir).st_uid
  if uid != os.geteuid():
    steve.die('you are not the effective owner of: %s', config.issue_dir)

  voters = steve.get_group(config.issue_dir + '/voters')
  if not voters:
    steve.die('"%s" must be an existing voter group: see votegroup', args.group)

  return voters


def create_issue_dir(issue_name, args, config):
  print 'Creating new issue:', issue_name

  # Note that .issue_dir already has the group.
  config.issue_dir += '/%s-%s' % (args.start, args.issue)
  if os.path.exists(config.issue_dir):
    steve.die('already exists: %s', config.issue_dir)
  os.mkdir(config.issue_dir, 0700)


def create_info_file(issue_name, args, config):
  info_fname = config.issue_dir + '/issue'

  contents = _use_template('templates/info-header.ezt', None, issue_name,
                           None, None, None, args, config)
  open(info_fname, 'w').write(contents)

  return info_fname


def build_hash(info_fname, voters, args):
  while True:
    issue_id = steve.filestuff(info_fname)
    monitors_hash = steve.get_hash_of('%s:%s' % (issue_id, args.monitors))

    hash = { }
    for voter in voters:
      h1 = steve.get_hash_of('%s:%s' % (issue_id, voter))
      h2 = steve.get_hash_of('%s:%s' % (issue_id, h1))

      hash[voter] = (h1, h2)
      hash[h1] = (voter, None)
      hash[h2] = (voter, None)

    if len(hash) == 3 * len(voters):
      return monitors_hash, hash

    _ = steve.get_input_line('anything to retry collision-free hash', True)
    open(info_fname, 'a').write('')


def create_monitors_file(args, config):
  monitors_fname = config.issue_dir + '/monitors'

  open(monitors_fname, 'w').write(args.monitors + '\n')

  return monitors_fname


def create_type_file(info_fname, args, config):
  type_fname = config.issue_dir + '/vote_type'

  f = open(type_fname, 'w')
  f.write('%s\n' % (args.votetype,))

  if args.selector == 0:
    ballots = ('yes', 'no', 'abstain')
  else:
    ballots = steve.ballots(open(info_fname).readlines())

  f.writelines(option + '\n' for option in ballots)

  return type_fname


def verify_email(issue_name, info_fname, args, config):
  print 'Here is the issue information to be sent to each voter:'

  contents = open(info_fname).read() \
             + _use_template('templates/explain.ezt', 'unique-hash-key', issue_name,
                             None, None, None, args, config)
  _basic_verify(contents, args, config)


def verify_voters(voters, args, config):
  print 'Here is the list of voter e-mail addresses:'

  contents = '\n'.join(voters)
  _basic_verify(contents, args, config)


def _basic_verify(contents, args, config):
  print '=============================================================='
  print contents
  print '=============================================================='

  # No need to stop and verify.
  if args.batch:
    return

  while True:
    answer = steve.get_input_line('"ok" to accept, or "abort" to delete issue', False)
    answer = answer.lower()
    if answer == 'abort':
      shutil.rmtree(config.issue_dir)
      sys.exit(1)
    if answer == 'ok':
      return


def email_monitors(issue_name, info_fname, monitors_hash, hash, args, config):
  # Collect hash signatures of voter files that should not change
  ### need to expand this. allow some to be missing during dev/test.
  sigs = ['%s: %s' % (steve.hash_file('make_issue.py'), 'make_issue.py'),
          ]

  msg = _use_template('templates/monitor-email.ezt', 'unique-hash-key',
                      issue_name, monitors_hash, hash, sigs,
                      args, config)

  ### mail the result. for now, print out what would have been mailed.
  print '### DEBUG'
  print msg
  print '###'


def email_voters():
  pass


def _use_template(template_fname, key, issue_name, monitors_hash, hash, sigs,
                  args, config):
  data = {
    'hashid': key,
    'type': args.votetype.strip('0123456789'),
    'issue_name': issue_name,
    'group': args.group,
    'selector': args.selector,
    'style': args.style,
    'hostname': config.hostname,
    'monitors': args.monitors,
    'monitors_hash': monitors_hash,
    'email': config.email,
    'file': args.file,
    'sigs': sigs,
    }
  if hash:
    data['count'] = len(hash)
    data['hash2'] = sorted(v[1] for v in hash.values() if v[1])

  buf = cStringIO.StringIO()
  ezt.Template(template_fname, compress_whitespace=False).generate(buf, data)
  return buf.getvalue()


if __name__ == '__main__':
  os.umask(0077)
  main()
